### PR TITLE
Adds "target pruning" for greater policy concentration & extra shine.

### PR DESCRIPTION
**Pull request by amj**
not yet ready for review, needs tests.

@brilee the generic 'reshape visit counts' would be a good place to consolidate the temperature squash/stretch logic, as well as any other hairbrained schemes, encapsulated behind nice flags.  What do you think?

### CODE REVIEW

File: cc/mcts_player.cc [hunk 1]
```diff
@@ -157,6 +159,10 @@ Coord MctsPlayer::SuggestMove(int new_readouts, bool inject_noise) {
     return Coord::kResign;
   }
 
+  if (options_.target_pruning && inject_noise) {
+    root_->ReshapeFinalVisits();
```
Comment (tommadams) [inline cc/mcts_player.cc]: this is the wrong place

File: cc/mcts_node.cc [hunk 2]
```diff
@@ -129,24 +129,52 @@ Coord MctsNode::GetMostVisitedMove() const {
...
+  MG_LOG(INFO) << "Best CAS: " << best_cas;
+  MG_LOG(INFO) << "Post_prune: ";
+  int total = 0;
+  float Q;
+  for (int i = 0; i < kNumMoves; ++i) {
+    // Store Q before modifications, as child_Q depends on edges[i].N.
+    Q = child_Q(i);
+    while ((Q * to_play + child_U(i) < best_cas) &&
+           (edges[i].N > 0)) {
+      edges[i].N --;
```
[the remark this answers is not in the document]
Comment (amj) [reply cc/mcts_node.cc]: Closed form implemented.

this is done once at the end of search in preparation for writing the training targets out.

File: cc/mcts_player.cc (hunk 1 above)
Comment (amj) [reply cc/mcts_player.cc]: Done.

**Comment by amj** [comment]
@tommadams would you mind reviewing the tests?

File: cc/mcts_node_test.cc [hunk 3]
```diff
@@ -428,6 +429,96 @@ TEST(MctsNodeTest, TestSelectLeaf) {
...
 }
 
+class ReshapeTargetTest : public ::testing::Test {
+ protected:
+  void SetUp() override {
+    best_ = -1;
+  }
+
+  void SearchTestablePosition(
+      TestablePosition p) {
```
[the remark this answers is not in the document]
Comment (amj) [reply cc/mcts_node_test.cc]: Done.

File: cc/mcts_node.cc [hunk 4]
```diff
@@ -129,24 +129,49 @@ Coord MctsNode::GetMostVisitedMove() const {
...
+    }
+
+    int new_N = std::max(0, std::min(int(child_N(i)),
+                         int(std::floor(-1 * (U_scale() * child_P(i) * std::sqrt(N())) /
+                         ( (child_Q(i) * to_play) - best_cas)) - 1 )));
+    total += edges[i].N - new_N;
+    edges[i].N = new_N;
+  }
+
+  MG_LOG(INFO) << "Pruned " << total << " visits.";
```
[the remark this answers is not in the document]
Comment (amj) [reply cc/mcts_node.cc]: Can i leave this in DEBUG?  i'd love to track this on stackdriver to get a sense of how often it's doing things.

File: cc/mcts_node.cc [hunk 5]
```diff
@@ -129,24 +129,49 @@ Coord MctsNode::GetMostVisitedMove() const {
...
+  float best_cas =
+      CalculateSingleMoveChildActionScore(to_play, U_common, uint16_t(best));
+
+  int total = 0;
+  for (int i = 0; i < kNumMoves; ++i) {
+    if (i == uint16_t(best)) {
+      continue;
+    }
+
+    int new_N = std::max(0, std::min(int(child_N(i)),
```
[the remark this answers is not in the document]
Comment (amj) [reply cc/mcts_node.cc]: Done.

File: cc/mcts_player.cc [hunk 6]
```diff
@@ -329,6 +340,8 @@ void MctsPlayer::UpdateGame(Coord c) {
     std::reverse(models.begin(), models.end());
   }
 
+  // Reshape the targets if needed.
```
[the remark this answers is not in the document]
Comment (amj) [reply cc/mcts_player.cc]: Done.

File: cc/mcts_node_test.cc [hunk 7]
```diff
@@ -13,6 +13,7 @@
 // limitations under the License.
 
 #include "cc/mcts_node.h"
+#include <sys/types.h>
```
[the remark this answers is not in the document]
Comment (amj) [reply cc/mcts_node_test.cc]: i don't even notice when stuff like this gets added.  my apologies!  changed to size_t

File: cc/mcts_node_test.cc [hunk 8]
```diff
@@ -428,6 +429,96 @@ TEST(MctsNodeTest, TestSelectLeaf) {
...
+      prob = 0.001;
+    }
+    probs[17] = 0.99;
+
+    MctsNode::EdgeStats root_stats;
+    root_ = new MctsNode(&root_stats, p);
+    root_->SelectLeaf()->IncorporateResults(0.0, probs, 0, root_);
+
+    MctsNode* leaf;
+    for (int i = 0; i < 10000; ++i) {
```
[the remark this answers is not in the document]
Comment (amj) [reply cc/mcts_node_test.cc]: Done.  I believe i've written it out coherently but do let me know if it makes sense.

File: cc/mcts_node.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (amj) [reply cc/mcts_node.cc]: Done; especially if we're planning on refactoring this out of the node it'd be good to be sure that the comments suffice.

File: cc/mcts_node.cc [hunk 9]
```diff
@@ -129,24 +129,51 @@ Coord MctsNode::GetMostVisitedMove() const {
...
+  float best_cas =
+      CalculateSingleMoveChildActionScore(to_play, U_common, uint16_t(best));
+
+  //int total = 0;
+  for (int i = 0; i < kNumMoves; ++i) {
+    if (i == uint16_t(best)) {
+      continue;
+    }
+
+    // Change N_child to the smallest value that satisfies the inequality
```
Comment (sethtroisi) [inline cc/mcts_node.cc]: add a note about why we do this.

"When we explored this position we had uncertainty about the value of this move,
now after we have expanded this node, we change the visit count to reflect how many
visits we would have given it with our updated understand of it's value (and the value of the best move)"

File: cc/mcts_player.h [hunk 10]
```diff
@@ -109,6 +109,10 @@ class MctsPlayer {
     float fastplay_frequency = 0;
     int fastplay_readouts = 20;
 
+    // Adjust the targets after reading to discard reads caused by 'unhelpful'
```
Comment (sethtroisi) [inline cc/mcts_player.h]: target pruning adjusts the targtes after reading...

**Pull request merged**